File: python/compare_python.py
from typing import List
import subprocess
import random

# FIXME: constants
FIND_N_CASES = 10
MY_PY_CODE = "./my.py"
ANSWER_PY_CODE = "./main.py"
OUT_FILE_NAME = "output.txt"
INPUT_PROPS = {
    'T': 10,
    'W': 5,
}

# FIXME: input generator
def make_input(props: dict) -> str:
    T, W = props['T'], props['W']
    random_T = random.randint(1, T)
    random_W = random.randint(1, W)
    args: List[int] = []

    for i in range(random_T):
        args.append(random.randint(1, 2))

    return f"{random_T} {random_W}\n" + "\n".join(str(arg) for arg in args)

################################################################################
################################################################################
################################################################################

class InputGenerator:
    def __init__(self, props: dict, generator=make_input):
        self.props = props
        self.generator = generator
        self.input = ""
        self.make()

    def make(self):
        self.input = self.generator(self.props)

class CodeProcessor:
    def __init__(self, code: str):
        self.code = code

    def execute(self, input_str: str) -> str:
        result = subprocess.run(['python3', self.code], input=input_str.encode(), stdout=subprocess.PIPE)
        return result.stdout.decode().strip()

class FindDifferentCases:
    def __init__(self, my_code: CodeProcessor, answer_code: CodeProcessor, input_gen: InputGenerator, finding_cases=FIND_N_CASES, out_file=OUT_FILE_NAME):
        self.my_code = my_code
        self.answer_code = answer_code
        self.input_gen = input_gen
        self.finding_cases = finding_cases
        self.out_file = out_file
        self.case = 0

    def run(self):
        self.make_file()

        while self.case < self.finding_cases:
            self.setup()
            if self.is_output_different():
                self.log_to_console()
                self.log_to_file()
                self.case += 1

    def setup(self):
        self.input_gen.make()
        self.my_code.execute(self.input_gen.input)
        self.answer_code.execute(self.input_gen.input)

    def is_output_different(self) -> bool:
        return self.my_result != self.answer_result

    def log_to_console(self):
        print(self.pretty_print)

    def log_to_file(self):
        with open(self.out_file, 'a') as f:
            f.write(self.pretty_print)

    def make_file(self):
        with open(self.out_file, 'w'):
            pass

    @property
    def my_result(self) -> str:
        return self.my_code.execute(self.input_gen.input)

    @property
    def answer_result(self) -> str:
        return self.answer_code.execute(self.input_gen.input)

    @property
    def pretty_print(self) -> str:
        return f"{self.input_gen.input}\nexpected: {self.answer_result}\nreceived: {self.my_result}\n\n"

def main():
    compare = FindDifferentCases(
        my_code=CodeProcessor(MY_PY_CODE),
        answer_code=CodeProcessor(ANSWER_PY_CODE),
        out_file=OUT_FILE_NAME,
        finding_cases=FIND_N_CASES,
        input_gen=InputGenerator(INPUT_PROPS)
    )
    compare.run()

if __name__ == '__main__':
    main()

File: typescript/compare_python.ts
// FIXME: constants
const FIND_N_CASES = 10;
const MY_PY_CODE = "./my.py";
const ANSWER_PY_CODE = "./main.py";
const OUT_FILE_NAME = "output.txt";
const INPUT_PROPS = {
  T: 10,
  W: 5,
};

// FIXME: input generator
function makeInput(props: InputGeneratorProps): string {
  const { T, W } = props;
  let randomT = Math.floor(Math.random() * T + 1);
  let randomW = Math.floor(Math.random() * W + 1);
  let args: number[] = [];

  for (let i = 0; i < randomT; i++) {
    args[i] = Math.floor(Math.random() * 2 + 1);
  }

  return [`${randomT} ${randomW}`, ...args].join("\n");
}

/************************************/
//   Below are not to be modified   //
/************************************/
// @ts-ignore - no type definition
import { spawnSync, execSync, SpawnSyncReturns } from "child_process";
// @ts-ignore - no type definition
import fs from "fs";
// @ts-ignore - no type definition
import { Buffer } from "node";

const COMPILER = "python3";
const FILE_EXTENSION = /\.py$/;

type InputGeneratorProps = typeof INPUT_PROPS;

interface FindDifferentCasesInterface {
  myCode?: CodeProcessor;
  answerCode?: CodeProcessor;
  inputGen?: InputGenerator;
  findingCases?: number;
  outFile?: string;
}

class InputGenerator {
  private props: InputGeneratorProps;
  private generator: (props: InputGeneratorProps) => string;
  input = "";

  constructor(props: InputGeneratorProps, generator = makeInput) {
    this.props = props;
    this.generator = generator;

    this.make();
  }

  make() {
    this.input = this.generator(this.props);
  }
}

class CodeProcessor {
  private code: string;

  constructor(code: string) {
    this.code = code;
  }

  execute(input: string) {
    const result = spawnSync("python3", [this.code], {
      input,
    });
    return result.stdout.toString().trim();
  }
}

class FindDifferentCases {
  private myCode: CodeProcessor;
  private answerCode: CodeProcessor;
  private inputGen: InputGenerator;
  private findingCases: number;
  private outFile: string;
  private case = 0;

  constructor({
    myCode = new CodeProcessor("my.cpp"),
    answerCode = new CodeProcessor("main.cpp"),
    inputGen = new InputGenerator(INPUT_PROPS),
    findingCases = FIND_N_CASES,
    outFile = "output.txt",
  }: FindDifferentCasesInterface) {
    this.myCode = myCode;
    this.answerCode = answerCode;
    this.inputGen = inputGen;
    this.findingCases = findingCases;
    this.outFile = outFile;
  }

  run() {
    this.makeFile();

    while (this.case < this.findingCases) {
      this.setup();
      if (this.isOutputDifferent()) {
        this.logToConsole();
        this.logToFile();
        this.case++;
      }
    }
  }

  private setup() {
    this.inputGen.make();
    this.myCode.execute(this.inputGen.input);
    this.answerCode.execute(this.inputGen.input);
  }

  private isOutputDifferent() {
    return this.myResult !== this.answerResult;
  }

  private logToConsole() {
    console.log(this.prettyPrint);
  }

  private logToFile() {
    fs.appendFileSync(this.outFile, this.prettyPrint);
  }

  private makeFile() {
    fs.writeFileSync(this.outFile, "");
  }

  private get myResult() {
    return this.myCode.execute(this.inputGen.input);
  }
  private get answerResult() {
    return this.answerCode.execute(this.inputGen.input);
  }
  private get prettyPrint() {
    return `${this.inputGen.input}\nexpected: ${this.answerResult}\nreceived: ${this.myResult}\n\n`;
  }
}

const main = () => {
  const compare = new FindDifferentCases({
    myCode: new CodeProcessor(MY_PY_CODE),
    answerCode: new CodeProcessor(ANSWER_PY_CODE),
    outFile: OUT_FILE_NAME,
    findingCases: FIND_N_CASES,
    inputGen: new InputGenerator(INPUT_PROPS),
  });
  compare.run();
};

main();
